### DISCUSSION TITLE
Why my prisma client always tries to connect to 127.0.0.1 in docker?

### DISCUSSION THREAD

**Discussion by rantiev** [prisma/prisma#20765] [Q&A] [answered]
### Question

This is docker config I have 
https://stackoverflow.com/questions/76940883/nestjs-cant-connect-to-mongo-while-running-in-docker 

Two containers are there: mongo container, and nestjs container. 

Nest.js (I assume prisma client by saying it) always tries to connect to 127.0.0.1 whatever I put to my scheme.prisma and env varialbes in .env file, or env variables in docker configs. 

I always regenerate prisma client with 'pnpm exec prisma generate'

### Expected behavior (optional)

Nestjs prisma client respect database url config IP address.

### Information about Prisma Schema, Client Queries and Environment (optional)
```prisma
<!-- Paste your Prisma schema below. Do not include your database credentials when sharing your Prisma schema! -->
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = "mongodb://root:prisma@mongo:27018/bookmeon?authSource=admin"
}

enum OrderStatus {
  pending
  confirmed
  completed
  cancelled
}
...
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @rantiev 👋 

In your docker config file, you are setting `MONGO_REPLICA_HOST` to `127.0.0.1`, which means that the MongoDB instance inside the container will try to connect to `127.0.0.1`, which refers to itself within the container. Can you set the `MONGO_REPLICA_HOST` to the service name of the MongoDB container within your Docker network?

i.e  `MONGO_REPLICA_HOST: mongo` instead of  `MONGO_REPLICA_HOST: 127.0.0.1`


If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by rantiev** [reply]
Hi, Thank you for the suggestion. But I have already tried this and it didn't work. The fact is prisa client somehow tries to connect to 127.0.0.1 even if I have configured it to connect to `mongo` (service name). It anyways tries to connect to 127. though real address is another I guess.

**Reply by rantiev** [reply]
So, I have tried again and it doesn't work.

**Reply by janpio** [reply]
Prisma does not have any special logic that could replace `mongo` with `127.0.0.1`. You give it a connection string, Prisma gives that to the Mongo Driver, and the Mongo Driver does its thing.

You did not include the error message here, but I assume it's this one from StackOverflow:

```
book_me_on-web-1  | PrismaClientInitializationError: Raw query failed. Code: `unknown`. Message: `Server selection timeout: No available servers. Topology: { Type: ReplicaSetNoPrimary, Servers: [ { Address: 127.0.0.1:27018, Type: Unknown, Error: Connection refused (os error 111) }, ] }`
book_me_on-web-1  |     at r (/my/node_modules/.pnpm/@prisma+client@5.0.0_prisma@5.0.0/node_modules/@prisma/client/runtime/library.js:103:2766)
book_me_on-web-1  |     at async Proxy.onModuleInit (/my/dist/prisma.service.js:14:9)
book_me_on-web-1  |     at async Promise.all (index 0)
book_me_on-web-1  |     at async callModuleInitHook (/my/node_modules/.pnpm/@nestjs+core@10.0.0_@nestjs+common@10.0.0_@nestjs+platform-express@10.0.0_reflect-metadata@0.1.13_rxjs@7.8.1/node_modules/@nestjs/core/hooks/on-module-init.hook.js:43:5)
book_me_on-web-1  |     at async NestApplication.callInitHook (/my/node_modules/.pnpm/@nestjs+core@10.0.0_@nestjs+common@10.0.0_@nestjs+platform-express@10.0.0_reflect-metadata@0.1.13_rxjs@7.8.1/node_modules/@nestjs/core/nest-application-context.js:223:13)
book_me_on-web-1  |     at async NestApplication.init (/my/node_modules/.pnpm/@nestjs+core@10.0.0_@nestjs+common@10.0.0_@nestjs+platform-express@10.0.0_reflect-metadata@0.1.13_rxjs@7.8.1/node_modules/@nestjs/core/nest-application.js:97:9)
book_me_on-web-1  |     at async NestApplication.listen (/my/node_modules/.pnpm/@nestjs+core@10.0.0_@nestjs+common@10.0.0_@nestjs+platform-express@10.0.0_reflect-metadata@0.1.13_rxjs@7.8.1/node_modules/@nestjs/core/nest-application.js:166:33)
book_me_on-web-1  |     at async bootstrap (/my/dist/main.js:21:5) {
book_me_on-web-1  |   clientVersion: '5.0.0',
book_me_on-web-1  |   errorCode: 'P2010'
book_me_on-web-1  | }
book_me_on-web-1  | 
book_me_on-web-1  | Node.js v20.5.1
```

The error message `Server selection timeout: No available servers. Topology: { Type: ReplicaSetNoPrimary, Servers: [ { Address: 127.0.0.1:27018, Type: Unknown, Error: Connection refused (os error 111) }, ] }` is reported by Prisma, but comes from the Mongo Rust Driver that is trying to work with your connection string.

I assume that Mongo somehow goes from `mongodb://root:prisma@mongo:27018/bookmeon?authSource=admin` to a server called `127.0.0.1:27018` that it tries to connect to. (MongoDB connection strings are just pointers, that tell the driver what to _actually_ connect to.) 

I don't know, but would not be surprised if this was via the `MONGO_REPLICA_HOST` and `MONGO_REPLICA_PORT` you are configuring. Can your for example set `MONGO_REPLICA_PORT: 99999` and see if that then pops up in your error message?

**Comment by rantiev** [comment, marked as answer]
I was sure that it's something with prisma client generation IP address substitution, but turned out it was my docker daemon. I have completely reinstalled docker daemon (removed all images/containers as well) and reinstalled all images, recreated containers again. 

And with no changes nestjs starts well now. Thanks to everyone!
